From 6fc4c1c0cd57fcb2dc490132a080f8bc33679d74 Mon Sep 17 00:00:00 2001
From: Callum <38501234+calumari@users.noreply.github.com>
Date: Fri, 7 Aug 2020 19:25:21 +0100
Subject: [PATCH] Tamable EAR immunity

Prevent tamable animals (cats, dogs and parrots) immune to ear IF they are tamed and not sitting. Helps prevent tamed mobs from getting separated from their owner.
---
 .../java/net/minecraft/server/EntityTameableAnimal.java     | 6 ++++++
 .../socialhangover/conduit/config/ConduitWorldConfig.java   | 5 +++++
 src/main/java/org/spigotmc/ActivationRange.java             | 1 +
 3 files changed, 12 insertions(+)

diff --git a/src/main/java/net/minecraft/server/EntityTameableAnimal.java b/src/main/java/net/minecraft/server/EntityTameableAnimal.java
index 9b0d937f06..5d2df57f72 100644
--- a/src/main/java/net/minecraft/server/EntityTameableAnimal.java
+++ b/src/main/java/net/minecraft/server/EntityTameableAnimal.java
@@ -184,4 +184,10 @@ public abstract class EntityTameableAnimal extends EntityAnimal {
     public void setWillSit(boolean flag) {
         this.willSit = flag;
     }
+
+    // Conduit start
+    public boolean shouldBeActive() {
+        return isTamed() && !isSitting();
+    }
+    // Conduit end
 }
diff --git a/src/main/java/net/socialhangover/conduit/config/ConduitWorldConfig.java b/src/main/java/net/socialhangover/conduit/config/ConduitWorldConfig.java
index 10bf2e8f39..19e01912aa 100644
--- a/src/main/java/net/socialhangover/conduit/config/ConduitWorldConfig.java
+++ b/src/main/java/net/socialhangover/conduit/config/ConduitWorldConfig.java
@@ -105,4 +105,9 @@ public class ConduitWorldConfig {
     private void sleepSettings() {
         tickInsomniaInCreative = getBoolean("sleep.tick-insomnia-in-creative", false);
     }
+
+    public boolean tamableImmuneToEAR;
+    private void tamableSettings() {
+        tamableImmuneToEAR = getBoolean("mobs.tamable.immune-to-EAR", false);
+    }
 }
diff --git a/src/main/java/org/spigotmc/ActivationRange.java b/src/main/java/org/spigotmc/ActivationRange.java
index b4356fc876..625cf32384 100644
--- a/src/main/java/org/spigotmc/ActivationRange.java
+++ b/src/main/java/org/spigotmc/ActivationRange.java
@@ -378,6 +378,7 @@ public class ActivationRange
      */
     public static boolean checkIfActive(Entity entity)
     {
+        if (entity.world.conduitConfig.tamableImmuneToEAR && entity instanceof net.minecraft.server.EntityTameableAnimal && ((net.minecraft.server.EntityTameableAnimal) entity).shouldBeActive()) return true; // Conduit
         // Never safe to skip fireworks or entities not yet added to chunk
         if ( !entity.inChunk || entity instanceof EntityFireworks ) {
             return true;
-- 
2.28.0.windows.1

